materialize-snowflake/sqlgen.go

```diff
@@ -212,6 +215,27 @@ SELECT * FROM (SELECT -1, CAST(NULL AS VARIANT) LIMIT 0) as nodoc
 {{ end -}}
 {{ end }}
 
+{{ define "pipe_name" -}}
```

Comment:
I think ideally we would have a single pipe among different shards, since that would allow optimal utilisation of the pipe's queue. The main issue here is that when we `CREATE OR REPLACE` a pipe, we will clobber up the pipe that might be used by another shard...

It is not possible to change the `COPY INTO` statement of a pipe, which means it is not possible to _update_ a pipe in response to changes in projections of a collection, we need to recreate the pipe.

Re-creating a pipe before we run `Acknowledge` would break the `Acknowledge` recovery phase, this means we can't create the pipe during `Apply` phase as we create the tables.

We are creating pipes during `Store` phase, and in this phase I don't think there is a way for us to ensure recreation of the pipe without risking clobbering some other shard's work. Let me know if you can think of a way to do this.

With that, we may have to create shard-specific pipes similar to what we do in Databricks 🤔

Reply:
I also think that ideally there'd be a single pipe among different shards. This concept of coordination could be optimal in other ways too, such as a standard updates materialization where there could be sharded "workers" that only serialize data and upload files and then the "leader" (the one with responsible for the key range starting at 0 perhaps) is the one that issues the `MERGE` queries, rather than each shard queuing up its separate, smaller merge queries on the same table concurrently.

I could imagine achieving this using a table in Snowflake to coordinate between shards, somehow...it would take some thought. For a standard updates materialization this would make more sense as it would involve waking up the warehouse and they'd do that anyway, but a purely snowpipe materialization not so much. I don't think I know enough about how existing transactions are coordinated between shards to say much more. I do recall that there is some element of them committing all around the same time. So maybe there's something we could do there, and use driver checkpoints in a clever way.

It would seem fine for now to use shard-specific pipes for the sake of getting this done (and sharded materializations are not really functional atm for other reasons), although could be something interesting to discuss more.